src/optics.jl

```diff
@@ -318,6 +323,185 @@ function _modify(f, obj, r::Recursive, ::ModifyBased)
     end
 end
 
+"""
+    modify_stateful(f, (obj,state), optic) => Tuple{NewValue,NewState}
+
+Here `f` has signature `f(::Value, ::State) => Tuple{NewValue,NewState}`.
+"""
+function modify_stateful end
+
+@generated function modify_stateful(f::F, (obj, state)::T, optic::Properties) where {T,F}
+    _modify_stateful_inner(T)
+end
+
+# Separated for testing object/state combinations without restarts
+function _modify_stateful_inner(::Type{<:Tuple{O,S}}) where {O,S}
+    modifications = []
+    vals = Expr(:tuple)
+    fns = fieldnames(O)
+    for (i, fn) in enumerate(fns)
+        v = Symbol("val$i")
+        st = if S <: ContextState 
+            if O <: Tuple 
+                :(ContextState(state.vals, obj, StaticInt{$(QuoteNode(fn))}()))
+            else
+                :(ContextState(state.vals, obj, StaticSymbol{$(QuoteNode(fn))}()))
+            end
+        else
+            :state
+        end
+        ms = :(($v, state) = f(getfield(props, $(QuoteNode(fn))), $st))
+        push!(modifications, ms)
+        push!(vals.args, v)
+    end
+    patch = O <: Tuple ? vals : :(NamedTuple{$fns}($vals))
+    start = :(props = getproperties(obj))
+    rest = MacroTools.@q begin
```

Comment:
just for my education, is there an advantage of MacroTools.@q over quote here?

Reply:
Maybe it has some effect on variable scope as well? I can't actually remember the reason now.